Allow Streams to be instantiated using GD resources

settermjd opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

I've been working on a patch to laminas-barcode, which requires a Stream to be able to be created from a GD resource. However, the patch fails, as Diactoros only allows resource types to be of type stream. So, this patch expands the list of types allowed to instantiate a Stream to include GD types. I'm submitting this after discussing the idea with @weierophinney some time ago. I hope that I understood what he meant.

settermjd committed:
Allow Streams to be instantiated using GD resources

I've been working on a patch to laminas-barcode to create a Stream from
a GD resource, but am unable to, as the resource type has to be a
stream. However, after talking to mwop, some time ago, he suggested that
it'd be fine to patch Diactoros to allow for GD resource types to
instantiate streams as well; hence, this patch.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

weierophinney commented:
Ping me for a review when it's ready, @settermjd !

Also, don't forget to use the `-s` switch to sign-off your commits! There's instructions in the contributors guide on how to use rebase to add them.

samsonasik commented on src/Stream.php:
```diff
@@ -357,4 +362,15 @@ private function setStream($stream, string $mode = 'r') : void
+     */
+    protected function isAllowedStreamResourceType($resource)
+    {
+        return in_array(get_resource_type($resource), self::ALLOWED_TYPES);
```
may need strict check : 

```suggestion
        return in_array(get_resource_type($resource), self::ALLOWED_TYPES, true);
```

settermjd replied:
Thanks for the feedback.

samsonasik replied:
you seems push to wrong repo branch, should push to your fork branch instead of this repo branch

samsonasik replied:
please push to your fork ( as this PR ):

https://github.com/settermjd/laminas-diactoros/tree/expand-list-of-resource-types

instead of this repo branch:

https://github.com/laminas/laminas-diactoros/tree/expand-list-of-resource-types

settermjd replied:
I must have done that by accident. Sorry.

settermjd replied:
That said, I updated the PR to include strict type checking.

settermjd committed:
Move test for valid stream resource type to a trait

As the same code was used in both Stream and StreamFactory, it, to me,
made sense to make it reusable by creating a trait to store it and then
reusing that code among Stream and StreamFactory.. Seeing that traits
are already used in the library, it seemed okay to create another one
for use in this context.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update the docs to reflect the change of allowed resource types

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Add strict typing when checking the resource's type

Thanks to feedback from @samsonasik.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd commented:
> Ping me for a review when it's ready, @settermjd !
> 
> Also, don't forget to use the `-s` switch to sign-off your commits! There's instructions in the contributors guide on how to use rebase to add them.

Ready for review. It's not a big patch, and I hope that I understood what you were saying in our previous discussion a little while ago.

samsonasik commented on src/StreamTrait.php:
```diff
@@ -0,0 +1,25 @@
+     * @param resource $resource Stream resource.
+     * @return bool
+     */
+    protected function isValidStreamResourceType($resource)
```
I think modifier can be `private` instead of `protected`

samsonasik commented on src/Stream.php:
```diff
@@ -40,6 +40,13 @@
+    /**
+     * A list of types that are allowed to instantiate a Stream
+     */
+    const ALLOWED_TYPES = ['stream', 'gd'];
```
modifier can be added, eg: private

settermjd replied:
I don't have an objection, but would that really be necessary?

settermjd replied:
Actually, it seems necessary to keep it public, as they're referenced in the call to `isValidStreamResourceType` in `StreamFactory`, on line 51.

samsonasik replied:
/cc @weierophinney

settermjd replied:
After a bit of thought, @samsonasik, I moved Stream::ALLOWED_TYPES to be a private variable in StreamTrait. Thanks for the suggestion and encouragement, and thanks to @weierophinney for further educating me.

settermjd committed:
Change StreamTrait::isValidStreamResourceType visibility to private

I don't expect the visibility to be anything other than private, so
setting it appropriately. The previous public visibility was an
oversight on my part.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Refactor Stream::ALLOWED_TYPES constant to StreamTrait

After quite a bit of thinking about this, I feel that @samsonasik is
correct. It doesn't make sense, to me, to expose an internal detail for
no good reason. Given that, this change moves the constant to the trait
and sets its visibility to private.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Sort list of allowed types in alphabetical order

Perhaps it's a minor thing, but @localheinz always pushed for
readability wherever possible. So, to make the code just that much more
readable is worth it; which is the motivation for this, minor, change.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

samsonasik commented on src/StreamTrait.php:
```diff
@@ -0,0 +1,30 @@
+     * @param resource $resource Stream resource.
+     * @return bool
+     */
+    private function isValidStreamResourceType($resource)
```
return declaration can be added `: bool` so no need `@return bool` docblock

settermjd replied:
Thanks for spotting that.

lcobucci commented on src/Stream.php:
```diff
@@ -40,6 +40,8 @@
  */
 class Stream implements StreamInterface
 {
+    use StreamTrait;
```
I'd suggest avoiding traits here.

I'd argue that only the stream should be responsible for the validation of its input.
I understand that the purpose of the patch is to extend the behaviour but I believe we could rethink this design.

settermjd replied:
I've been pondering on this for a bit. Seems to make sense to me. Will refactor.

lcobucci commented on src/StreamTrait.php:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * A list of types that are allowed to instantiate a Stream
+     */
+    private $allowedTypes = ['gd', 'stream'];
```
This should rather be a private constant (which works great with the centralisation of the validation on the stream class).

Let's also remember that properties on traits are replicated in every instance, which might be a not so great usage of memory.

settermjd replied:
@lcobucci, I took your advice and rolled the functionality from StreamTrait into Stream. Thanks for your guidance.

settermjd committed:
Refactor StreamTrait functionality into Stream.php

Following on from the advice of @lcobucci, this change refactors the
StreamTrait functionality into Stream, and removes the trait from both
Stream and StreamFactory. On thinking about it, it makes most sense to
have this functionality directly in Stream, and not, in effect,
duplicated in multiple locations by another form.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

lcobucci commented on src/StreamFactory.php:
```diff
@@ -21,6 +21,7 @@
 
 class StreamFactory implements StreamFactoryInterface
 {
+
```
I know this is just coding standard, but since it hasn't been reported by CI... based on the rest of the codebase it looks like this shouldn't be added.

settermjd replied:
I'll check this out. Thanks for letting me know.

settermjd replied:
Fixed!

lcobucci reviewed (approved):
Just a teeny tiny thing! Thanks for your work here... things look much better and the stream class is much more coherent to me :+1:

lcobucci commented:
The failing integration test seems to come from `php-http/psr7-integration-tests`... it's doing external calls without retrying on failure, which makes the suite a bit flaky.

settermjd committed:
Remove space breaking code style checks

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

weierophinney reviewed (approved):
:ship:

settermjd commented:
Big thanks to everyone for all your support on this PR.
